MdeModulePkg/Include/Protocol/NonDiscoverableDeviceUniqueId.h

```diff
@@ -0,0 +1,67 @@
+  For example, if a device registers with the NonDiscoverableDeviceRegistrationLib,
+  they can also publish this new protocol on the same EFI_HANDLE as RegisterNonDiscoverableMmioDevice.
+  NonDiscoverablePciDeviceDxe will consume this UniqueId if found and set it to the device's UniqueId.
+  If not found, will back up to the prior method of assigning the UniqueId via a static counter.
+  The platform must maintain all UniqueId's assigned are unique for every instance of the protocol.
+
+  The platform may need to use this new protocol defined UniqueId instead because
+  the current UniqueId that gets used in NonDiscoverablePciDeviceDxe is a static counter,
+  which can lead to non-deterministic value's being returned by PciIo->GetLocation().
+  In order to have a deterministic UniqueId returned by GetLocation() this protocol
+  can be used to specify a specific UniqueId for a given Handle.
+
+  The uniqud ID must be in the range of:
+    MAX_NON_DISCOVERABLE_PCI_DEVICE_ID/2 <= UniqueId < MAX_NON_DISCOVERABLE_PCI_DEVICE_ID
+    MAX_NON_DISCOVERABLE_PCI_DEVICE_ID == 32 * 256
+
+  Usage example:
+    EFI_HANDLE Handle = NULL;
+
+    Status = RegisterNonDiscoverableMmioDevice (
+                NonDiscoverableDeviceTypeAhci,
+                NonDiscoverableDeviceDmaTypeCoherent,
+                NULL,
+                &Handle,
+                1,
+                x, y
+              );
+
+      NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL * UniqueIdProtocol = (NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL*) AllocateZeroPool (sizeof (NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL));
+      if (UniqueIdProtocol == NULL) {
+        return EFI_OUT_OF_RESOURCES;
+      }
+      UniqueIdProtocol->Revision = NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL_REVISION;
+      UniqueIdProtocol->UniqueId = 0x1000; // MAX_NON_DISCOVERABLE_PCI_DEVICE_ID/2 <= UniqueId < MAX_NON_DISCOVERABLE_PCI_DEVICE_ID
+
+      Status = gBS->InstallMultipleProtocolInterfaces (&Handle, &gEdkiiNonDiscoverableDeviceUniqueIdProtocolGuid, UniqueIdProtocol, NULL);
+
+  Copyright (c) Microsoft Corporation.
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL_H
+#define NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL_H
+
+#include <Uefi/UefiBaseType.h>
+
+#define EDKII_NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL_GUID \
+  { 0xd60d0e74, 0x5c4e, 0x4d45, { 0x9f, 0xe0, 0x2a, 0x3e, 0x4f, 0x9d, 0x55, 0xc1 } }
+
+#define NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL_REVISION  1
+
+// We will use the upper half of the ID space for this UnqueId protocol assignment.
+// The lower half can be used as a backup by NonDiscoverablePciDeviceDxe when the protocol is not available.
+// The uniqud ID must be in the range of: MAX_NON_DISCOVERABLE_PCI_DEVICE_ID/2 <= UniqueId < MAX_NON_DISCOVERABLE_PCI_DEVICE_ID
+#define MAX_NON_DISCOVERABLE_PCI_DEVICE_ID  (32 * 256)
+
+typedef struct _NON_DISCOVERABLE_DEVICE_UNIQUE_ID_PROTOCOL {
+  UINTN    Revision;

```

Comment:
nit: I think this is clearer, but feel free to close.

```suggestion
  UINTN    ProtocolRevision;
```